concordium-std/src/impls.rs

```diff
@@ -251,46 +391,1124 @@ impl Write for ContractState {
         if self.current_position.checked_add(len).is_none() {
             return Err(());
         }
-        let num_bytes = unsafe { write_state(buf.as_ptr(), len, self.current_position) };
+        let num_bytes = unsafe {
+            prims::state_entry_write(self.state_entry_id, buf.as_ptr(), len, self.current_position)
+        };
+        if num_bytes == u32::MAX {
+            return Err(()); // Entry did not exist.
+        }
         self.current_position += num_bytes; // safe because of check above that len + pos is small enough
         Ok(num_bytes as usize)
     }
 }
 
-impl HasContractState<()> for ContractState {
-    type ContractStateData = ();
+impl<StateApi: HasStateApi> VacantEntryRaw<StateApi> {
+    /// Create a new `VacantEntryRaw`.
+    pub(crate) fn new(key: Key, state_api: StateApi) -> Self {
+        Self {
+            key,
+            state_api,
+        }
+    }
 
+    /// Gets a reference to the key that would be used when inserting a value
+    /// through the `VacantEntryRaw`.
     #[inline(always)]
-    fn open(_: Self::ContractStateData) -> Self {
-        ContractState {
-            current_position: 0,
+    pub fn key(&self) -> &[u8] { &self.key }
+
+    /// Sets the value of the entry with the `VacantEntryRaw`’s key.
+    pub fn insert(mut self, value: &[u8]) -> Result<StateApi::EntryType, StateError> {
+        let mut entry = self.state_api.create_entry(&self.key)?;
+        entry.write_all(value).unwrap_abort(); // Writing to state cannot fail.
+        entry.move_to_start(); // Reset cursor.
+        Ok(entry)
+    }
+
+    /// Sets the value of the entry with the `VacantEntryRaw`’s key.
+    /// This differs from
+    /// [`insert`](Self::insert) in that it automatically serializes
+    /// the provided value. [`insert`](Self::insert) should be preferred
+    /// for values that can be directly converted to byte arrays, e.g., any
+    /// value that implements [`AsRef<[u8]>](AsRef).
+    pub fn insert_serial<V: Serial>(
```

Comment:
We use `x` for the auto-serializing version and `x_raw` for the byte version elsewhere. Perhaps we should rename this to match?

Reply:
Yeah, makes sense.